lk2nd/smb1360/smb1360-qcom.c

```diff
@@ -15,21 +22,105 @@ static const struct smb1360_battery batteryB = {
 	.profile = SMB1360_BATTERY_PROFILE_B,
 };
 
+static const struct smb1360_battery batteryK = {
+	.name = "Default Profile",
+};
+
+static int64_t abs64(int64_t val)
+{
+	return val < 0 ? (-1 * val) : val;
+}
+
+static int smb1360_check_batt_profile(const void *fdt, int offset, uint32_t connected_rid)
+{
+	const uint32_t *prop;
+	status_t ret;
+
+	int i, len, timeout, loaded_profile, new_profile;
+	uint8_t val = 0;
+	uint32_t profile_rid[BATTERY_PROFILE_MAX], rid10;
+
+	prop = fdt_getprop(fdt, offset, "qcom,profile-a-rid-kohm", &len);
+	if (len < 0)
+		return -EINVAL;
+	profile_rid[0] = fdt32_to_cpu(*prop);
+
+	prop = fdt_getprop(fdt, offset, "qcom,profile-b-rid-kohm", &len);
+	if (len < 0)
+		return -EINVAL;
+	profile_rid[1] = fdt32_to_cpu(*prop);
+
+	rid10 = connected_rid / 10;
+
+	if ((profile_rid[0] - connected_rid) < rid10) {
+		new_profile = SMB1360_BATTERY_PROFILE_A;
+	} else if ((profile_rid[1] - connected_rid) < rid10) {
+		new_profile = SMB1360_BATTERY_PROFILE_B;
+	} else {
+		new_profile = SMB1360_BATTERY_PROFILE_KEEP;
+	}
+
+	return new_profile;
+}
+
+static uint32_t smb1360_parse_batt_id(const void *fdt, int offset)
+{
+	const uint32_t *prop;
+	int64_t denom, batt_id_uv;
+	int64_t denom_dividend, denom_quotient, connected_rid_dividend, connected_rid_quotient;
+	int32_t len, volt, rpull = 0, vref = 0;
+	uint32_t connected_rid;
+
+	prop = fdt_getprop(fdt, offset, "qcom,batt-id-vref-uv", &len);
+	if (len == sizeof(vref))
+		return -EINVAL;
+	vref = fdt32_to_cpu(*prop);
+
+	prop = fdt_getprop(fdt, offset, "qcom,batt-id-rpullup-kohm", &len);
+	if (len != sizeof(rpull))
+		return -EINVAL;
+	rpull = fdt32_to_cpu(*prop);
+
+	volt = pm8x41_adc_channel_read(VADC_BAT_CHAN_ID);
+	batt_id_uv = volt * 1000;
+
+	if (batt_id_uv == 0)
+		return -EINVAL;
+
+	denom_dividend = vref * 1000000LL;
+	denom_quotient = batt_id_uv;
+	denom = (abs64(denom_dividend) / abs64(denom_quotient)) - 1000000LL;
```

Comment:
Hmm, downstream uses `denom = div64_s64(vref * 1000000LL, batt_id_uv) - 1000000LL;`, so a 64-bit signed division. Where does the abs64 come from?

Reply:
I was looking into `div64_s64` to see how it performed the division and saw it there, https://elixir.bootlin.com/linux/v4.9/source/lib/div64.c#L163. By checking the results, downstream returns a value for `connected_rid` of 100. Without `abs` I get -99, with `abs` I get 99. I guess the difference of 1 in result is given by some round. What do you think?

Reply:
If I'm reading this correctly `div64_s64` does an unsigned 64-bit division with abs first and then in the `t = (dividend ^ divisor) >> 63;` line computes the sign separately and adds it back. So effectively the result should be the one without abs(). It seems like something else is odd here...

Reply:
Ohh...I'll rework on it then, thank you